File: ruby/test/statement_printer_test.rb
require "minitest/autorun"
require "json"
require_relative "../lib/theatrical_plays/statement_printer"

class TestStatementPrinter < Minitest::Test
  def test_statement_with_known_types
    invoice = load_json_fixture('invoice.json')
    plays = load_json_fixture('plays.json')

    expected_statement = "Statement for BigCo\n" \
    " Hamlet: $650.00 (55 seats)\n" \
    " As You Like It: $580.00 (35 seats)\n" \
    " Othello: $500.00 (40 seats)\n" \
    "Amount owed is $1730.00\n" \
    "You earned 47 credits\n"

    assert_equal expected_statement, TheatricalPlays::StatementPrinter.new.print(invoice, plays)
  end

  def test_statement_with_new_types
    invoice = load_json_fixture('invoice_new_plays.json')
    plays = load_json_fixture('new_plays.json')

    error = assert_raises(StandardError) do
      TheatricalPlays::StatementPrinter.new.print(invoice, plays)
    end

    assert_equal 'unknown type: history', error.message
  end

  private

  def load_json_fixture(filename)
    file_path = File.join(__dir__, 'data', filename)
    JSON.parse(File.read(file_path))
  end
end

File: dart/test/statement_printer_test.dart
import 'package:test/test.dart';
import 'package:theatrical_plays/invoice.dart';
import 'package:theatrical_plays/performance.dart';
import 'package:theatrical_plays/play.dart';
import 'package:theatrical_plays/statement_printer.dart';

void main() {
  test('example statement', () {
    final plays = {
      'hamlet': Play('Hamlet', 'tragedy'),
      'as-like': Play('As You Like It', 'comedy'),
      'othello': Play('Othello', 'tragedy'),
    };

    final invoice = Invoice('BigCo', [
      Performance('hamlet', 55),
      Performance('as-like', 35),
      Performance('othello', 40),
    ]);

    final statementPrinter = StatementPrinter();
    final result = statementPrinter.print(invoice, plays);

    const expectedResult = 'Statement for BigCo\n'
        '  Hamlet: \$650.00 (55 seats)\n'
        '  As You Like It: \$580.00 (35 seats)\n'
        '  Othello: \$500.00 (40 seats)\n'
        'Amount owed is \$1,730.00\n'
        'You earned 47 credits\n';

    expect(result, expectedResult);
  });

  test('statement with new play types throws error', () {
    final plays = {
      'henry-v': Play('Henry V', 'history'),
      'as-like': Play('As You Like It', 'pastoral'),
    };

    final invoice = Invoice('BigCo', [
      Performance('henry-v', 53),
      Performance('as-like', 55),
    ]);

    final statementPrinter = StatementPrinter();
    
    expect(
      () => statementPrinter.print(invoice, plays),
      throwsA(
        isA<Exception>().having(
          (e) => e.toString(),
          'message',
          contains('unknown type: '))));
  });
}